packages/dapi/lib/chainDataProvider/ChainDataProvider.js

```diff
@@ -0,0 +1,155 @@
+const { EventEmitter } = require('events');
+const { BlockHeader, ChainLock } = require('@dashevo/dashcore-lib');
+const log = require('../log');
+
+/**
+ * Data access layer with caching support
+ */
+class ChainDataProvider extends EventEmitter {
+  /**
+   *
+   * @param coreRpcClient {CoreRpcClient}
+   * @param zmqClient {ZmqClient}
+   * @param blockHeadersCache {BlockHeadersCache}
+   */
+  constructor(coreRpcClient, zmqClient, blockHeadersCache) {
+    super();
+
+    this.coreRpcAPI = coreRpcClient;
+    this.zmqClient = zmqClient;
+    this.blockHeadersCache = blockHeadersCache;
+
+    this.chainLock = null;
+  }
+
+  /**
+   * @private
+   * @param blockHash {string}
+   */
+  blockHashHandler(blockHash) {
+    this.emit(this.events.NEW_BLOCK_HEADER, blockHash);
+  }
+
+  /**
+   * @private
+   * @param rawChainLock {string}
+   */
+  chainLockHandler(rawChainLock) {
+    this.chainLock = new ChainLock(Buffer.from(rawChainLock, 'hex'));
+
+    this.emit(this.events.NEW_CHAIN_LOCK);
+  }
+
+  /**
+   * Grabs most recent chainlock
+   * @returns {Promise<void>}
+   */
+  async init() {
+    let chainLock;
+
+    try {
+      chainLock = await this.coreRpcAPI.getBestChainLock();
+    } catch (e) {
+      if (e.code === -32603) {
+        log.info('No chain lock available in dashcore node');
+      } else {
+        throw e;
+      }
+    }
+
+    this.chainLock = new ChainLock(chainLock);
+
+    this.zmqClient.on(this.zmqClient.topics.rawtx, this.chainLockHandler);
+    this.zmqClient.on(this.zmqClient.topics.rawblock, this.blockHashHandler);
+  }
+
+  /**
+   * Get block hash by height
+   * @param height {number}
+   * @returns {Promise<string>}
+   */
+  async getBlockHash(height) {
+    return this.coreRpcAPI.getBlockHash(height);
+  }
+
+  /**
+   * Get block header by block hash
+   * @param blockHash {string}
+   * @returns {Promise<BlockHeader>}
+   */
+  async getBlockHeader(blockHash) {
+    const cached = this.blockHeadersCache.get(blockHash);
+
+    if (cached) {
+      return new BlockHeader(cached);
+    }
+
+    const rawBlockHeader = await this.coreRpcAPI.getBlockHeader(blockHash);
+    const blockHeaderBuffer = Buffer.from(rawBlockHeader, 'hex');
+
+    this.blockHeadersCache.set(blockHash, blockHeaderBuffer);
```

Comment:
I would recommend to store instances of classes, so you don't need to parse blobs everytime.

Reply:
Just realized that we also need to handle reorgs. So I wouldn't cache blocks that are not chain locked or last six if chainlock is not present or lower.